internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+use std::{cmp, collections::HashMap, fmt};
+
+// PolyType represents a generic parametrized type.
+//
+// TODO:
+//     Do not derive PartialEq implementation.
+//     Instead provide a custom implementation
+//     that instantiates both polytypes with the
+//     same type variables.
+//
+//     Note this requires a substitution, so remove
+//     this derivation once substitutions are defined.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct PolyType {
+    pub free: Vec<Tvar>,
+    pub bnds: Option<HashMap<Tvar, Kind>>,
```

Comment:
Can a Tvar have multiple Kinds? Meaning is this valid? 

```
forall ['a] where 'a is Addable and 'a is Comparable (x:'a) -> 'a
```

If so I think we need a different structure here. Would a list of tvar kind pairs be sufficient?

Reply:
At the moment as well the foreseeable future type variables will only be bound by a single kind. Having said that, a list might be better.

Reply:
What kind constraints does this code have?

```
change = (x,y) => (x - y) / (x + y)
```

I would expect a type signature like this:

```
forall ['a] where 'a is Addable and 'a is Subtractable and 'a is Divisible (x:'a, y:'a) -> 'a
```

Reply:
In this case the type signature would be:
```
forall ['a] where 'a is Divisible (x:'a, y:'a) -> 'a
```
as `Divisible` is the smallest class satisfying the above constraints.

Reply:
Ok, that makes sense. Does the logic about which Kind is smallest get applied during unification? Meaning that when we see Addable unify with Divisible we have logic that knows that they do unify but the resulting kind is the Divisible one? Not sure I am asking that clearly, in other words where does the logic about which Kind is the smallest satisfying kind live?

Reply:
You're being quite clear :). This logic is not present at the moment but will live with the kind itself. So each kind will know how to "unify" with other kinds.

Reply:
Sounds good. Thanks!